File: py/libs/socketservice/socketclient.py
# coding: utf8
# socketclient.py
# jichi 4/28/2014

if __name__ == '__main__':
  import sys
  sys.path.append('..')

__all__ = ['SocketClient']

from PySide.QtCore import QObject, Signal, QByteArray
#from PySide.QtNetwork import QAbstractSocket
from sakurakit.skclass import Q_Q
from sakurakit.skdebug import dprint
import socketmarshal, socketprotocol

class SocketClient(QObject):

  def __init__(self, parent=None):
    super(SocketClient, self).__init__(parent)
    self.__d = _SocketClient(self)

  connected = Signal()
  disconnected = Signal()
  socketError = Signal()

  dataReceived = Signal(bytearray) # data

  def sendData(self, data):  # str -> bool
    return self.__d.send(data)

  def address(self): return self.__d.address # -> str
  def setAddress(self, v): self.__d.address = v

  def port(self): return self.__d.port # -> int
  def setPort(self, v): self.__d.port = v

  def isActive(self): # -> bool, is connected
    s = self.__d.socket
    return bool(s) and s.state() == s.ConnectedState

  def isReady(self): # -> bool, is connected or disconnected instead of connecting or uninitialized
    s = self.__d.socket
    return bool(s) and s.state() in (s.ConnectedState, s.UnconnectedState)

  def start(self): self.__d.start()
  def stop(self): self.__d.stop()

  def waitForReady(self):
    s = self.__d.socket
    if s and s.state() not in (s.ConnectedState, s.UnconnectedState):
      from PySide.QtCore import QEventLoop
      loop = QEventLoop()
      s.stateChanged.connect(loop.quit)
      s.error.connect(loop.quit)
      loop.exec_();
      while s.state() in (s.HostLookupState, s.ConnectingState):
        loop.exec_()

@Q_Q
class _SocketClient(object):
  def __init__(self, q):
    self.address = '127.0.0.1' # host name without http prefix
    self.port = 0 # int
    self.socket = None # # QTcpSocket

  def _createSocket(self):
    from PySide.QtNetwork import QTcpSocket
    q = self.q
    ret = QTcpSocket(q)
    ret.error.connect(q.socketError)
    ret.connected.connect(q.connected)
    ret.disconnected.connect(q.disconnected)
    return ret

  def start(self):
    from PySide.QtNetwork import QHostAddress
    if not self.socket:
      self.socket = self._createSocket()
    self.socket.connectToHost(QHostAddress(self.address), self.port)
    dprint("pass")

  def stop(self):
    if self.socket and self.socket.isOpen():
      self.socket.close()
      dprint("pass")

  def send(self, data): # -> bool
    ok = False
    # Explicitly use QByteArray to preserve message size
    if not isinstance(data, QByteArray):
      data = QByteArray(data)
    size = len(data) + socketprotocol.MESSAGE_HEADER_SIZE
    header = socketmarshal.int32bytes(size)
    data.prepend(header)
    #assert len(data) == size
    if self.socket:
      ok = len(data) == self.socket.write(data)
    dprint("pass: ok = %s" % ok)
    return ok

if __name__ == '__main__':
  import sys
  from PySide.QtCore import QCoreApplication
  app =  QCoreApplication(sys.argv)
  c = SocketClient()
  c.setPort(6002)
  c.start()
  c.waitForReady()
  c.sendData("hello")
  c.disconnected.connect(app.quit)

  sys.exit(app.exec_())

# EOF

File: cpp/libs/qtsocketsvc/socketclient.cc
// socketclient.cc
// 4/29/2014 jichi
#include "qtsocketsvc/socketdef.h"
#include "qtsocketsvc/socketclient.h"
#include "qtsocketsvc/socketio_p.h"
#include <QtCore/QEventLoop>
#include <QtNetwork/QHostAddress>
#include <QtNetwork/QTcpSocket>

#define DEBUG "socketclient"
#include <sakurakit/skdebug.h>

//QTSS_BEGIN_NAMESPACE

/** Private class */

class SocketClientPrivate
{
  SK_DECLARE_PUBLIC(SocketClient)
public:
  QTcpSocket *socket;
  int port;
  QString address;
  quint32 currentDataSize; // current message body size read from socket

  explicit SocketClientPrivate(Q *q)
    : q_(q), socket(nullptr), port(0), address(SOCKET_SERVICE_HOST), currentDataSize(0) {}

  void createSocket()
  {
    socket = new QTcpSocket(q_);
    q_->connect(socket, SIGNAL(readyRead()), SLOT(readSocket()));

    q_->connect(socket, SIGNAL(error(QAbstractSocket::SocketError)), SIGNAL(socketError()));
    q_->connect(socket, SIGNAL(connected()), SIGNAL(connected()));
    q_->connect(socket, SIGNAL(disconnected()), SIGNAL(disconnected()));
  }

  bool writeSocket(const QByteArray &data);
  QByteArray readSocket();

  void dumpSocketInfo() const; // for debug only
};


bool SocketClientPrivate::writeSocket(const QByteArray &data)
{ return socket && SocketService::writeSocket(socket, data); }

QByteArray SocketClientPrivate::readSocket()
{
  if (Q_UNLIKELY(!socket))
    return QByteArray();
  return SocketService::readSocket(socket, currentDataSize);
}

void SocketClientPrivate::dumpSocketInfo() const
{
  if (socket)
    DOUT("socket"
         ": localAddress ="<< socket->localAddress() <<
         ", localPort ="   << socket->localPort() <<
         ", peerAddress =" << socket->peerAddress() <<
         ", peerPort ="    << socket->peerPort() <<
         ", state ="       << socket->state() <<
         ", error ="       << socket->errorString());
  else
    DOUT("socket = null");
}

/** Public class */

// Constructions:

SocketClient::SocketClient(QObject *parent)
  : Base(parent), d_(new D(this))
{}

SocketClient::~SocketClient() { delete d_; }

int SocketClient::port() const { return d_->port; }
void SocketClient::setPort(int value) { d_->port = value; }

QString SocketClient::address() const { return d_->address; }
void SocketClient::setAddress(const QString &value) { d_->address = value; }

void SocketClient::start()
{
  if (!d_->socket)
    d_->createSocket();
  d_->socket->connectToHost(QHostAddress(d_->address), d_->port);
  DOUT("pass");
}

void SocketClient::stop()
{
  if (d_->socket && d_->socket->isOpen()) {
    d_->socket->close();
    DOUT("pass");
  }
}

bool SocketClient::isActive() const
{
  return d_->socket && d_->socket->state() == QAbstractSocket::ConnectedState;
}

bool SocketClient::isReady() const
{
  return d_->socket && (
    d_->socket->state() == QAbstractSocket::ConnectedState ||
    d_->socket->state() == QAbstractSocket::UnconnectedState
  );
}

void SocketClient::waitForReady()
{
  if (d_->socket &&
    d_->socket->state() != QAbstractSocket::ConnectedState &&
    d_->socket->state() != QAbstractSocket::UnconnectedState) {

    QEventLoop loop;
    connect(d_->socket, SIGNAL(stateChanged(QAbstractSocket::SocketState)), &loop, SLOT(quit()));
    connect(d_->socket, SIGNAL(error(QAbstractSocket::SocketError)), &loop, SLOT(quit()));
    do loop.exec();
    while (!d_->socket ||
           d_->socket->state() == QAbstractSocket::HostLookupState ||
           d_->socket->state() == QAbstractSocket::ConnectingState);
  }
}

// I/O:

bool SocketClient::sendData(const QByteArray &data, int waitTime)
{
  bool ok = d_->writeSocket(data);
  if (ok && waitTime)
    ok = d_->socket->waitForBytesWritten(waitTime);
  return ok;
}

bool SocketClient::waitForConnected(int interval)
{ return d_->socket && d_->socket->waitForConnected(interval); }

bool SocketClient::waitForDisconnected(int interval)
{ return d_->socket && d_->socket->waitForDisconnected(interval); }

bool SocketClient::waitForBytesWritten(int interval)
{ return d_->socket && d_->socket->waitForBytesWritten(interval); }

bool SocketClient::waitForReadyRead(int interval)
{ return d_->socket && d_->socket->waitForReadyRead(interval); }

void SocketClient::readSocket()
{
  if (Q_LIKELY(d_->socket))
    while (d_->socket->bytesAvailable()) {
      QByteArray data = d_->readSocket();
      if (data.isEmpty())
        break;
      else
        emit dataReceived(data);
    }
}

void SocketClient::dumpSocketInfo() const { d_->dumpSocketInfo(); }

//QTSS_END_NAMESPACE

// EOF
